[WIP] UI, DataTable - Actions

nhaagen opened the pull request:
Based on https://github.com/ILIAS-eLearning/ILIAS/pull/2679, this is an idea on Actions in tables.

Please refer to the [docs](https://github.com/ILIAS-eLearning/ILIAS/pull/2724/files#diff-488841fa74a3b3894b10afa09f10f507R209-R246) and the [example](https://github.com/ILIAS-eLearning/ILIAS/pull/2724/files#diff-da67b86bfcde546a13848a188b8970a2).

Please ask and comment ;)

nhaagen committed:
UI, DataTable - Actions

  - promote Actions to Component, add to factory
  - implementation of Actions for single and selected rows;
  - relay action-parameters (row-ids) via JS
  - basic col-implementation

Amstutz commented:
Hi @nhaagen

Thanks for this proposal. I really like seeing this coming to live. Thx for the example, that did help me a lot. Neat to also offer a concept for multi actions.

Some Questions:

- [x]  Scope Argument: Why do you propose having the type of actions as an optional argument in the factory method? IMO this has two drawbacks: 

1. It is implicit, one has to read to docs to be aware of the different type of actions. 
2. It is implemented as optional argument. Recently I try to avoid such in the factory methods, since it makes it rather hard to add a mandatory lather. 

Both issues would be avoided if we just add the three types of actions through factory: 
```
$f->table()->action()->standard('Delete', 'ids', $url) OR  
$f->table()->action()->multi('Delete', 'ids', $url) OR  
$f->table()->action()->single('Delete', 'ids', $url)
```

Probably revise:
- [x]  Aria Labels: I believe we need aria labels for actions. Should we make them mandatory or can we guess them through the label by default?
- [x]  withDisabledAction: If I understand correctly, this is an opt-out for the „single" and both „actions". So, if for some reason one line does not offer „delete“, they can opt out here. Correct? If so please clarify by commenting the „withDisabledAction“ Method somewhat. Maybe also drop a sentence in the actions composition description (that there is an option to opt out for rows).

Looking forward 
Timon

nhaagen commented:
Hi @Amstutz ,
thanks for your comments. I removed the Scope-attribute from Actions in favor of a full factory and added a line on "withDisabledAction".

As for Aria Labels - Actions will end up to be Buttons (maybe in a Dropdown) and might just pass an aria down. Or could explicitly set an aria-label derived from the Action's label and some record-attribute. I think that the label itself should suffice, but if we go there I'd opt for the second option with a very sophisticated aria-label.

Cheers, 
Nils

Amstutz commented:
Hi @nhaagen 

Thanks a lot, looks good to me.

@klees This is fine for JF discussion from my point of you. Push it to the agenda as soon as good for you.

Thx.

matthiaskunkel commented:
Jour Fixe, 08 JUN 2020 : Richard presented the concept for actions in the new UI Table and asks for comments and questions until the next JF. Once, an implementation (HTML) of the new table is available, it would be good to present it to the SIG Accessibility for input and for asking if additional area labels are necessary or if the labels in the actions itself are sufficient.

mjansenDatabay commented:
:shipit:

matthiaskunkel commented:
Jour Fixe, 22 JUN 2020 : We highly appreciate this suggestion and accept the PR for trunk. Please merge.

Amstutz commented:
@nhaagen I wait with merging for the implementation part, I hope this is fine with you.

klees commented on src/UI/Component/Table/Action/Action.php:
```diff
@@ -0,0 +1,19 @@
+
+interface Action extends \ILIAS\UI\Component\Component
+{
+
```
```suggestion
```

klees commented on src/UI/Component/Table/Action/Action.php:
```diff
@@ -0,0 +1,19 @@
+    const VALID_TARGET_CLASSES = [Signal::class, URI::class];
+
+    public function getLabel() : string;
+    public function getParameterName() : string;
```
```suggestion

    public function getParameterName() : string;
```

klees commented on src/UI/Component/Table/DataRetrieval.php:
```diff
@@ -18,4 +18,5 @@ public function getRows(
         array $visible_column_ids,
         array $additional_parameters
     ) : \Generator;
+
```
```suggestion
```

klees commented on src/UI/Implementation/Component/Table/Action/Factory.php:
```diff
@@ -0,0 +1,23 @@
+    public function standard(string $label, string $parameter_name, $target) : I\Standard
+    {
+        return new Standard($label, $parameter_name, $target);
+    }
```
```suggestion
    }

```

klees commented on src/UI/Implementation/Component/Table/Action/Factory.php:
```diff
@@ -0,0 +1,23 @@
+    public function single(string $label, string $parameter_name, $target) : I\Single
+    {
+        return new Single($label, $parameter_name, $target);
+    }
```
```suggestion
    }

```

klees reviewed (changes requested):
Hi @nhaagen!

This looks quite functional and also decently modularized. I have some ideas for better modularization and also some other requests and questions. Thanks for adding "declare(strict_type=1)", btw. I would be happy to keep the size of the PR manageable and this was already a lot to digest. So feel free to ask to postpone stuff and try not to add more code. I'm definitely happy if don't look into the `ViewControls` in this PR. As always, ping me if we need to discuss some of my items in person.

So let's start with questions:

- [ ] **(Aria-)Role**: I checked the accessibilty-rules quite thoroughly, and they seem to implemented as demanded. I wondered if they are phrased as required, though. Is it really only `role`, not `aria-role` which we want as accessibility attribute? Same goes for the other a11y-attributes that are requested. Please double check and also amend the rules accordingly, if required.
* [ ] **Column Type**: I'm somehow missing a switch for the column type in the Renderer. Currently it looks as if all cells will have the same formatting or (worse) as if the consumer of the service will need to format it's stuff to match the columns type. Am I missing a check somewhere? Wouldn't we need to switch according to column type [in the `Renderer` for StandardRow](https://github.com/ILIAS-eLearning/ILIAS/pull/2724/files?file-filters%5B%5D=.js&file-filters%5B%5D=.php#diff-9b8a11e17baf37cba16b17a1d70e346bR282) to enforce the same rendering accross a column? Would we need to check the data that consumers pass to the `RowFactory` to do that reliably? I think we should put a little more thought into the architecture there. Maybe we could use the `Refinery` and the information from the columns to derive a reusable check for the data that users provide us in the `RowFactory`. And/or add some methods to accomplish the rendering and checking tasks to the columns.

...go one with the suggestions:

* [ ] **Page Size**: Why do you use two different names (`$number_of_rows` vs. `$page_size`) [here](https://github.com/ILIAS-eLearning/ILIAS/pull/2724/files#diff-3e946421ac306127678b306505b21f4cR50)? If semantics don't change there (I think they don't) please use the same name for the property as for the argument to the constructor.
* [ ] **Treeshake**: Why don't we [use treeshake](https://github.com/ILIAS-eLearning/ILIAS/pull/2724/files#diff-d8041c83348f31e52db230c4be1cd78aR7) when packaging our JS?
- [ ] **Type Enumeration**: Why do you [enumerate the column types in the base class?](https://github.com/ILIAS-eLearning/ILIAS/pull/2724/files?file-filters%5B%5D=.js&file-filters%5B%5D=.php#diff-e2859d84dd137e9d38baf7e8c54a90adR13). This is odd, as we do not use that enumeration anywhere (I may be missing something) and also have other ways to deal with hierarchies like that (abstract methods, match on class name, visitor pattern, ...). I think 

...and finish with the change requests:

- [ ] **Dropdowns Only**: Always put the actions in Dropdowns, even if there is only one action available. If Buttons and Dropdowns are mixed in one table the look is, hmm, wild. I don't think we should concern ourselfs with that, at least now.
- [ ] **Tests**: Add them, they are missing for PHP and JS. I can even bear more code for them =) Please try write explicits tests for the `Renderer` on method level in an explicit class, not just for the HTML in the general component test. The `Renderer` seems to be complex enough to justify that and therefore could also use some factorization, which will be required for the tests as well.
- [ ] and the minor inline stuff.

I'm really exited to see a good plan to be executed so nicely =)

Thanks!

nhaagen committed:
do not enumerate colum-types (use classes instead), align var-name 'page_size/number of rows', always render dropdown in favor of single buttons

nhaagen committed:
remove/add a few (blank lines)
Co-authored-by: Richard Klees <richard.klees@concepts-and-training.de>

nhaagen committed:
UI/Table/Data:
 - move a bunch of rendering functions into component
 - only yield single actions to row(factory)
 - some tests for Table-classes

nhaagen committed:
UI/Table/Data:
- modulize param-handling
- add some JS-tests
- add Hint/Roadmap-entry to not construct RowFactories in the component-class
- remove selectable-flag from row

nhaagen committed:
use URI-functions to get rid of param-handling in Actions

nhaagen committed:
UI/Table/Data: polished rules a bit

nhaagen committed:
UI/Table/Data: fix rendering and add tests

nhaagen commented:
Hi @klees ,
I still owe you some answers:
- [x] **(Aria-)Role** I'm quite positive that it's just "role", not "aria-role"; however, it's still an aria-attribute.
I have orientated mainly on https://www.w3.org/TR/wai-aria-practices-1.1/#gridAndTableProperties.
I do not think that we want to include row-count and  -index at this point; there is a big fat [warning](https://www.w3.org/TR/wai-aria-practices-1.1/#gridAndTableProperties_rows) concerning screen-reader reactions on faulty indizes, and maybe the need for it is not too high when there is a well-formed  (in terms of a11y) pagination-control. However, open for discussion. 
Other aria-attributes are in.
- [x] **Column Type:** The cols area actually just in for the example's sake and to see how things fit togehter.
There is no actual switch/tranform yet. I'd go for Renderer::renderStandardRow or Row::getCellContent.
I can add it now or later...

With the present state I think the other issues are adressed:
- var-names for number_of_rows (changed)
- Treeshake (was inactive) 
- Type Enumeration (gotten rid of)
- Dropdowns (done)
- And - there are tests ;)

So, thanks again for digging through,
Nils

klees commented:
Hi @nhaagen,

I'm fine with elaborating column later on. I'm also fine with the current a11y-rules, if someone would want to change these we have a well defined process to do so.

Thanks for adding tests. I think we can move on with this by merging.

@Amstutz: Any objections to merge this?

Best regards!

Amstutz reviewed (changes requested):
Hi @nhaagen

Thanks a lot for this huge contribution. It took me some time to work through all this. There are several gems hidden in there. Special thanks for adding JS-Tests again. Nice to see this growing. However, we need a better place to store those. For the moment this is fine, but as mentioned last time, there is some ground work we still need to cover such as integrating those to CI and finding a good place for them to happily live and grow. Note for accessibility I checked, if I reach all interactive parts of the table by tabbing, this seemed to work. Lighthouse pointed to some minor issues, see bellow.


My Questions:
- [ ] RowFactory: Why the Roadmap entry instead of just proposing the factory where it probably should be? The placement of RowFactory is somewhat strange and hard to find. Why is this not offered through the main factories just like Action and Column? Is it because it is only used when injected by the table, so the consumer would never have to access it through through $DIC->ui()->….?
- [ ] The current layout those not look somewhat funky (lots of borders). Any hope @ckrahl takes a look at this before merging?
- [ ] In the last JF we discussed of Modals rather being part of other components than used as stand alone. You do not need to add this here, but to clarify, would it be possible to add something like „withDeletableEntries“ on the table, directly adding the InterruptiveModal for deleting those entries in the (near) future?
- [ ] NYI: What is the story behind „NYI“ in „withAdditionalViewControl“ and „withRequest“ in Data?



Please do or shortly point out why not:
- [ ] withIndex / getIndex Columns: When would I need those on the public interface? Maybe move to internal interface.
- [ ] getCellContent of Row: Why is this method on the public interface, isn’t that only used internally for rendering? Maybe move to internal interface.

Please do:
- [ ] Accessibility Lighthouse I: "Assistive technologies, like screen readers, can't interpret ARIA attributes with invalid values. Failing Elements table.table“ I guess, it is the empty aria-colcount of the Table.
- [ ]  Accessibility Lighthouse II: "Some ARIA parent roles must contain specific child roles to perform their intended accessibility functions (tr.header.row).“ Note the role is rowgroup here, but it does not contain any rows.  
- [ ] Sortation: It does not work yet, it always dislplays „none“.
- [ ] Action getParameterName, getTarget: Add some short description of what what param name is here. Same with Target. In my case, I was sure only after looking at the example.
- [ ] Examples I: There have been complaints in the past, that the examples are contradicting parts of the rules of the components used. Showing a lightbox modal after a multi action might be considered somewhat strange here. Beste change to e.g. a roundtrip modal with some content like „Comparing the selected IDs“, or further Actions with selected IDs or something along those lines.
- [ ] Examples II: Currently this huge component is somewhat weak on the documentation through examples part. There is just one covering basically everything. Add at least one short example for each Column and Action type. This could also help featuring the different nature of the Action. Also this could be used to show case some really easy tables focusing on one single aspect only.

Thanks a lot, looking forward to see this in action.

Amstutz commented:
For whoever might be reading along, the work here is not dead. It got somewhat pushed behind the work for releasing 7, but will take up steam again after that. We will keep you posted.

Amstutz commented:
This has been open for almost a year now. This is too long. We will re-open/create a new PR as soon as we continue our work here. We will continue on this track, but we are all atm burried in releasing ILIAS 7.
